File: brain/router.cpp
#include "brain/router.hpp"

#include <algorithm>
#include <regex>

#include "brain/tool_definitions.hpp"

namespace brain {
namespace {
const std::vector<std::string> kTimePhrases = {"what time", "what's the time", "current time", "what day is it", "what's the date", "what date", "may gio", "bay gio la may gio", "hom nay ngay may"};
const std::vector<std::string> kWeatherPhrases = {"weather in", "weather for", "what's the weather", "how's the weather", "temperature in", "weather now", "weather today", "thoi tiet", "nhiet do", "thoi tiet o", "hom nay troi"};
const std::vector<std::string> kNewsPhrases = {"news", "headlines", "what's happening", "whats happening", "current events", "top stories", "tin tuc", "thoi su", "tin moi"};
const std::vector<std::string> kSystemPhrases = {"system status", "how are you doing", "how are you feeling", "your temperature", "cpu temp", "health check", "how's your health", "how you doing", "tinh trang he thong", "trang thai he thong", "suc khoe he thong"};
const std::vector<std::string> kJokePhrases = {"tell me a joke", "joke", "make me laugh", "something funny", "say something funny", "ke chuyen cuoi", "ke mot cau dua", "lam toi cuoi"};
const std::vector<std::string> kLocalPhrases = {"hello", "hi", "hey", "good morning", "good afternoon", "good evening", "how are you", "what's up", "who are you", "what are you", "what's your name", "thank you", "thanks", "bye", "goodbye", "see you", "good night", "help", "what can you do", "xin chao", "chao", "ban la ai", "cam on", "tam biet", "ban khoe khong"};
}  // namespace

Router::Router(OllamaClient client) : client_(std::move(client)) {}

std::string Router::ToLower(std::string value) {
  std::transform(value.begin(), value.end(), value.begin(), [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return value;
}

bool Router::IsLocalChat(const std::string& user_input) const {
  const std::string user_lower = ToLower(user_input);
  for (const auto& phrase : kLocalPhrases) {
    if (user_lower.find(phrase) != std::string::npos) return true;
  }
  size_t words = 0;
  bool in_word = false;
  for (char c : user_lower) {
    if (std::isspace(static_cast<unsigned char>(c))) {
      in_word = false;
    } else if (!in_word) {
      in_word = true;
      ++words;
    }
  }
  return words <= 3 && user_input.find('?') == std::string::npos;
}

std::string Router::ExtractNewsCategory(const std::string& user_input) const {
  const std::string lower = ToLower(user_input);
  if (lower.find("tech") != std::string::npos) return "technology";
  if (lower.find("cong nghe") != std::string::npos) return "technology";
  if (lower.find("sport") != std::string::npos) return "sports";
  if (lower.find("the thao") != std::string::npos) return "sports";
  if (lower.find("medical") != std::string::npos) return "health";
  if (lower.find("suc khoe") != std::string::npos) return "health";
  if (lower.find("kinh doanh") != std::string::npos) return "business";
  if (lower.find("giai tri") != std::string::npos) return "entertainment";
  if (lower.find("khoa hoc") != std::string::npos) return "science";
  for (const auto& c : {"business", "entertainment", "health", "science", "sports", "technology"}) {
    if (lower.find(c) != std::string::npos) return c;
  }
  return "";
}

std::string Router::ExtractLocation(const std::string& user_input, const std::string& response_text) const {
  std::smatch m;
  std::regex response_pattern(R"(location["\s=:]+["']*([^"'\\]\s,]+))", std::regex::icase);
  if (std::regex_search(response_text, m, response_pattern) && m.size() > 1) return m[1].str();

  const std::vector<std::regex> patterns = {
      std::regex(R"(weather (?:in|for|at) ([A-Za-z\s]+))"),
      std::regex(R"(in ([A-Za-z]+))"),
      std::regex(R"(thoi tiet (?:o|tai) ([\w\s]+))"),
      std::regex(R"(([A-Z][a-z]+(?:\s[A-Z][a-z]+)*))")};
  for (const auto& p : patterns) {
    if (std::regex_search(user_input, m, p) && m.size() > 1) {
      std::string loc = m[1].str();
      const std::string lower = ToLower(loc);
      if (lower != "the" && lower != "is" && lower != "it" && lower != "what" && lower != "how" && lower != "like") return loc;
    }
  }
  return "";
}

std::pair<ToolType, nlohmann::json> Router::DetectToolFromText(const std::string& user_input, const std::string& response_text) const {
  const std::string user_lower = ToLower(user_input);
  const std::string response_lower = ToLower(response_text);
  if (response_lower.find("get_current_time") != std::string::npos) return {ToolType::kTime, nlohmann::json::object()};
  if (response_lower.find("get_weather") != std::string::npos) return {ToolType::kWeather, {{"location", ExtractLocation(user_input, response_text)}}};
  if (response_lower.find("get_news") != std::string::npos) return {ToolType::kNews, {{"category", ExtractNewsCategory(user_input)}}};
  if (response_lower.find("get_system_status") != std::string::npos) return {ToolType::kSystemStatus, nlohmann::json::object()};
  if (response_lower.find("get_joke") != std::string::npos) return {ToolType::kJoke, nlohmann::json::object()};
  if (response_lower.find("cloud_handoff") != std::string::npos) return {ToolType::kCloud, {{"query", user_input}}};

  for (const auto& p : kTimePhrases) if (user_lower.find(p) != std::string::npos) return {ToolType::kTime, nlohmann::json::object()};
  for (const auto& p : kWeatherPhrases) if (user_lower.find(p) != std::string::npos) return {ToolType::kWeather, {{"location", ExtractLocation(user_input, "")}}};
  for (const auto& p : kNewsPhrases) if (user_lower.find(p) != std::string::npos) return {ToolType::kNews, {{"category", ExtractNewsCategory(user_input)}}};
  for (const auto& p : kJokePhrases) if (user_lower.find(p) != std::string::npos) return {ToolType::kJoke, nlohmann::json::object()};
  for (const auto& p : kSystemPhrases) if (user_lower.find(p) != std::string::npos) return {ToolType::kSystemStatus, nlohmann::json::object()};
  if (IsLocalChat(user_input)) return {ToolType::kNone, nlohmann::json::object()};
  return {ToolType::kCloud, {{"query", user_input}}};
}

RouterResult Router::Route(const std::string& user_input) {
  nlohmann::json messages = nlohmann::json::array();
  messages.push_back({{"role", "system"}, {"content", SystemPrompt()}});
  const size_t start = conversation_history_.size() > 8 ? (conversation_history_.size() - 8) : 0;
  for (size_t i = start; i < conversation_history_.size(); ++i) messages.push_back(conversation_history_[i]);
  messages.push_back({{"role", "user"}, {"content", user_input}});

  const ChatResponse response = client_.Chat(messages, &Tools());
  conversation_history_.push_back({{"role", "user"}, {"content", user_input}});

  if (response.is_tool_call && !response.tool_calls.empty()) {
    const auto& call = response.tool_calls.front();
    ToolType t = ToolType::kNone;
    if (call.name == "get_current_time") t = ToolType::kTime;
    else if (call.name == "get_weather") t = ToolType::kWeather;
    else if (call.name == "get_news") t = ToolType::kNews;
    else if (call.name == "get_system_status") t = ToolType::kSystemStatus;
    else if (call.name == "get_joke") t = ToolType::kJoke;
    else if (call.name == "cloud_handoff") t = ToolType::kCloud;
    RouterResult result;
    result.tool = t;
    result.response = "";
    result.arguments = call.arguments;
    return result;
  }

  auto [tool, args] = DetectToolFromText(user_input, response.content);
  if (tool == ToolType::kNone) {
    conversation_history_.push_back({{"role", "assistant"}, {"content", response.content}});
    RouterResult result;
    result.tool = ToolType::kNone;
    result.response = response.content;
    result.arguments = nlohmann::json::object();
    return result;
  }
  RouterResult result;
  result.tool = tool;
  result.response = "";
  result.arguments = args;
  return result;
}

void Router::ClearHistory() {
  conversation_history_ = nlohmann::json::array();
}

}  // namespace brain

File: brain/router.py
"""
Main routing logic - single LLM for routing and chat.
Includes text-based tool detection fallback for smaller models.
"""

import re
from typing import Optional, Tuple
from dataclasses import dataclass
from enum import Enum

from .ollama_client import OllamaClient
from .tool_definitions import TOOLS, SYSTEM_PROMPT


class ToolType(Enum):
    TIME = "get_current_time"
    WEATHER = "get_weather"
    NEWS = "get_news"
    SYSTEM_STATUS = "get_system_status"
    JOKE = "get_joke"
    CLOUD = "cloud_handoff"
    NONE = "none"  # Direct chat response


@dataclass
class RouterResult:
    """Result from the router."""
    tool: ToolType
    response: Optional[str]  # Direct response if no tool
    arguments: dict  # Tool arguments if tool called


class Router:
    """Routes user queries to appropriate handlers."""

    # Keywords for text-based tool detection (using word boundary matching)
    TIME_PHRASES = [
        "what time", "what's the time", "current time", "what day is it", "what's the date", "what date",
        "mấy giờ", "may gio", "bây giờ là mấy giờ", "bay gio la may gio", "hôm nay ngày mấy", "hom nay ngay may",
    ]
    WEATHER_PHRASES = [
        "weather in", "weather for", "what's the weather", "how's the weather", "temperature in", "weather now", "weather today",
        "thời tiết", "thoi tiet", "nhiệt độ", "nhiet do", "thời tiết ở", "thoi tiet o", "hôm nay trời", "hom nay troi",
    ]
    NEWS_PHRASES = [
        "news", "headlines", "what's happening", "whats happening", "current events", "top stories",
        "tin tức", "tin tuc", "thời sự", "thoi su", "tin mới", "tin moi",
    ]
    SYSTEM_PHRASES = [
        "system status", "how are you doing", "how are you feeling", "your temperature", "cpu temp", "health check", "how's your health", "how you doing",
        "tình trạng hệ thống", "tinh trang he thong", "trạng thái hệ thống", "trang thai he thong", "sức khỏe hệ thống", "suc khoe he thong",
    ]
    JOKE_PHRASES = [
        "tell me a joke", "joke", "make me laugh", "something funny", "say something funny",
        "kể chuyện cười", "ke chuyen cuoi", "kể một câu đùa", "ke mot cau dua", "làm tôi cười", "lam toi cuoi",
    ]

    # Phrases that the local model can handle — simple chat, greetings, identity
    LOCAL_PHRASES = [
        "hello", "hi", "hey", "good morning", "good afternoon", "good evening",
        "how are you", "what's up", "who are you", "what are you", "what's your name",
        "thank you", "thanks", "bye", "goodbye", "see you", "good night",
        "help", "what can you do",
        "xin chào", "xin chao", "chào", "chao", "bạn là ai", "ban la ai",
        "cảm ơn", "cam on", "tạm biệt", "tam biet", "bạn khỏe không", "ban khoe khong",
    ]

    def __init__(self, ollama_client: OllamaClient):
        self.client = ollama_client
        self.conversation_history = []

    def _is_local_chat(self, user_input: str) -> bool:
        """Check if the input is simple enough for the local model."""
        user_lower = user_input.lower().strip()
        # Short greetings / simple chat
        for phrase in self.LOCAL_PHRASES:
            if phrase in user_lower:
                return True
        # Very short inputs (1-3 words) that aren't questions are likely greetings
        words = user_lower.split()
        if len(words) <= 3 and "?" not in user_input:
            return True
        return False

    def _extract_news_category(self, user_input: str) -> str:
        """Extract news category from user input."""
        user_lower = user_input.lower()
        categories = ["business", "entertainment", "health", "science", "sports", "technology"]
        # Also match common synonyms
        synonyms = {
            "tech": "technology",
            "sport": "sports",
            "medical": "health",
            "công nghệ": "technology",
            "cong nghe": "technology",
            "thể thao": "sports",
            "the thao": "sports",
            "sức khỏe": "health",
            "suc khoe": "health",
            "kinh doanh": "business",
            "giải trí": "entertainment",
            "giai tri": "entertainment",
            "khoa học": "science",
            "khoa hoc": "science",
        }
        for synonym, category in synonyms.items():
            if synonym in user_lower:
                return category
        for cat in categories:
            if cat in user_lower:
                return cat
        return ""

    def _detect_tool_from_text(self, user_input: str, response_text: str) -> Tuple[ToolType, dict]:
        """
        Detect tool from user input keywords and/or model response text.
        Fallback for models that don't use structured tool calls.
        """
        user_lower = user_input.lower()
        response_lower = (response_text or "").lower()

        # Priority 1: Check for tool mentions in model response (e.g., "[get_current_time]")
        if "get_current_time" in response_lower:
            return ToolType.TIME, {}

        if "get_weather" in response_lower:
            location = self._extract_location(user_input, response_text)
            return ToolType.WEATHER, {"location": location}

        if "get_news" in response_lower:
            category = self._extract_news_category(user_input)
            return ToolType.NEWS, {"category": category}

        if "get_system_status" in response_lower:
            return ToolType.SYSTEM_STATUS, {}

        if "get_joke" in response_lower:
            return ToolType.JOKE, {}

        if "cloud_handoff" in response_lower:
            return ToolType.CLOUD, {"query": user_input}

        # Priority 2: Check for specific phrases in user input
        for phrase in self.TIME_PHRASES:
            if phrase in user_lower:
                return ToolType.TIME, {}

        for phrase in self.WEATHER_PHRASES:
            if phrase in user_lower:
                location = self._extract_location(user_input, "")
                return ToolType.WEATHER, {"location": location}

        for phrase in self.NEWS_PHRASES:
            if phrase in user_lower:
                category = self._extract_news_category(user_input)
                return ToolType.NEWS, {"category": category}

        for phrase in self.JOKE_PHRASES:
            if phrase in user_lower:
                return ToolType.JOKE, {}

        for phrase in self.SYSTEM_PHRASES:
            if phrase in user_lower:
                return ToolType.SYSTEM_STATUS, {}

        # Priority 3: If it's simple chat, keep local
        if self._is_local_chat(user_input):
            return ToolType.NONE, {}

        # Priority 4: Everything else → cloud handoff
        # The local 1.5B model can't reliably answer knowledge/technical questions
        return ToolType.CLOUD, {"query": user_input}

    def _extract_location(self, user_input: str, response_text: str) -> str:
        """Extract location from user input or response."""
        # Try to find location in model response
        match = re.search(r'location["\s=:]+["\']*([^"\'\]\s,]+)', response_text, re.IGNORECASE)
        if match:
            return match.group(1)

        # Try common patterns in user input
        patterns = [
            r"weather (?:in|for|at) ([A-Za-z\s]+)",
            r"in ([A-Za-z]+)",
            r"thời tiết (?:ở|tại) ([\w\s]+)",
            r"thoi tiet (?:o|tai) ([\w\s]+)",
            r"([A-Z][a-z]+(?:\s[A-Z][a-z]+)*)"  # Capitalized words
        ]

        for pattern in patterns:
            match = re.search(pattern, user_input)
            if match:
                loc = match.group(1).strip()
                # Filter out common words
                if loc.lower() not in ["the", "is", "it", "what", "how", "like"]:
                    return loc

        return ""  # No location found, orchestrator will use config default

    def route(self, user_input: str) -> RouterResult:
        """
        Route user input to appropriate handler.
        """
        # Build messages
        messages = [
            {"role": "system", "content": SYSTEM_PROMPT}
        ]

        # Add conversation history (keep last 4 exchanges for context)
        messages.extend(self.conversation_history[-8:])

        # Add current user message
        messages.append({"role": "user", "content": user_input})

        # Get response with tool calling
        response = self.client.chat(messages, tools=TOOLS)

        # Process response
        if response.is_tool_call:
            # Model used structured tool calling
            tool_call = response.tool_calls[0]

            try:
                tool_type = ToolType(tool_call.name)
            except ValueError:
                tool_type = ToolType.NONE

            # Guardrail: small models can occasionally pick the wrong tool.
            # If the user input strongly matches a known tool keyword, prefer that.
            heuristic_tool, heuristic_args = self._detect_tool_from_text(user_input, "")
            if heuristic_tool != tool_type:
                # Prefer strong heuristic when it disagrees with a tool call.
                # - If heuristic suggests a concrete local tool → override.
                # - If heuristic suggests cloud handoff and model chose a local tool → override.
                if heuristic_tool not in (ToolType.NONE, ToolType.CLOUD):
                    tool_type = heuristic_tool
                    arguments = heuristic_args
                elif heuristic_tool == ToolType.CLOUD and tool_type not in (ToolType.CLOUD, ToolType.NONE):
                    tool_type = ToolType.CLOUD
                    arguments = {"query": user_input}
                else:
                    arguments = tool_call.arguments
            else:
                arguments = tool_call.arguments

            self.conversation_history.append(
                {"role": "user", "content": user_input}
            )

            return RouterResult(
                tool=tool_type,
                response=None,
                arguments=arguments
            )
        else:
            # Fallback: detect tool from text
            tool_type, arguments = self._detect_tool_from_text(
                user_input, response.content
            )

            self.conversation_history.append(
                {"role": "user", "content": user_input}
            )

            if tool_type == ToolType.NONE:
                # Simple chat — use the local model's response
                self.conversation_history.append(
                    {"role": "assistant", "content": response.content or ""}
                )
                return RouterResult(
                    tool=ToolType.NONE,
                    response=response.content,
                    arguments={}
                )
            else:
                # Tool or cloud handoff
                return RouterResult(
                    tool=tool_type,
                    response=None,
                    arguments=arguments
                )

    def clear_history(self):
        """Clear conversation history."""
        self.conversation_history = []
